src/codec.rs

```diff
@@ -0,0 +1,531 @@
+pub enum CodecError {
+    #[error("I/O error")]
+    Io(#[from] std::io::Error),
+    #[error("{0} bytes left in buffer after decoding value")]
+    BytesLeftOver(usize),
+    #[error("length prefix of encoded vector overflows buffer: {0}")]
+    LengthPrefixTooBig(usize),
+    #[error("other error: {0}")]
+    Other(#[source] Box<dyn Error + 'static + Send + Sync>),
+}
+
+/// Describes how to decode an object from a byte sequence.
+pub trait Decode<D>: Sized {
+    /// Read and decode an encoded object from `bytes`. `decoding_parameter` provides details of the
+    /// wire encoding such as lengths of different portions of the message. On success, the decoded
+    /// value is returned and `bytes` is advanced by the encoded size of the value. On failure, an
+    /// error is returned and no further attempt to read from `bytes` should be made.
+    fn decode(decoding_parameter: &D, bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError>;
+
+    /// Convenience method to get decoded value. Returns an error if [`Self::decode`] fails, or if
+    /// there are any bytes left in `bytes` after decoding a value.
+    fn get_decoded(decoding_parameter: &D, bytes: &[u8]) -> Result<Self, CodecError> {
+        let mut cursor = Cursor::new(bytes);
+        let decoded = Self::decode(decoding_parameter, &mut cursor)?;
+        if cursor.position() as usize != bytes.len() {
+            return Err(CodecError::BytesLeftOver(
+                bytes.len() - cursor.position() as usize,
+            ));
+        }
+
+        Ok(decoded)
+    }
+}
+
+/// Describes how to encode objects into a byte sequence.
+pub trait Encode {
+    /// Append the encoded form of this object to the end of `bytes`, growing the vector as needed.
+    fn encode(&self, bytes: &mut Vec<u8>);
+
+    /// Convenience method to get encoded value.
+    fn get_encoded(&self) -> Vec<u8> {
+        let mut ret = Vec::new();
+        self.encode(&mut ret);
+        ret
+    }
+}
+
+// This enables VDAFs to use () as an aggregation parameter.
+impl Decode<()> for () {
+    fn decode(_decoding_parameter: &(), _bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(())
+    }
+}
+
+// This enables VDAFs to use () as an aggregation parameter.
+impl Encode for () {
+    fn encode(&self, _bytes: &mut Vec<u8>) {}
+}
+
+impl Decode<()> for u8 {
+    fn decode(_decoding_parameter: &(), bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let mut value = [0u8; size_of::<u8>()];
+        bytes.read_exact(&mut value)?;
+        Ok(value[0])
+    }
+}
+
+impl Encode for u8 {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.push(*self);
+    }
+}
+
+impl Decode<()> for u16 {
+    fn decode(_decoding_parameter: &(), bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(bytes.read_u16::<BigEndian>()?)
+    }
+}
+
+impl Encode for u16 {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.extend_from_slice(&u16::to_be_bytes(*self));
```

Comment:
See [this comment](https://github.com/abetterinternet/libprio-rs/pull/166#discussion_r806371588) for rationale of not using `byteorder` here.

Reply:
This is unfortunate: `write_u16` and friends fail (IIUC) iff `write_all` fails; `write_all` fails on a subset of `write` failures; and [`Vec::write` will never fail, but this is not documented](https://doc.rust-lang.org/src/std/io/impls.rs.html#373-376), so it would be unwise to rely on this behavior.